File: python_system/python_shared/api/get_ads.py
"""
get_ads.py - Hybrid Database System
Fetches ads from SQLite database
Converted from PHP to Python
"""

from fastapi import APIRouter, Depends, Query
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
from sqlalchemy import desc
from typing import Optional
import json

import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).parent.parent))

from models import Ad, Category, Company
from database import get_db

router = APIRouter()


@router.get("/ads")
async def get_ads(
    page: int = Query(1, ge=1),
    q: str = Query("", description="Search query"),
    category: str = Query("", description="Category filter"),
    company: str = Query("", description="Company filter"),
    sort: str = Query("date", description="Sort by: date, views, favs, ai"),
    pageSize: int = Query(12, ge=1, le=100),
    db: Session = Depends(get_db)
):
    """Fetch ads with filters and pagination"""
    try:
        # Build query
        query = db.query(Ad).filter(Ad.status == "active")

        # Company filter
        if company:
            query = query.filter(Ad.company_slug == company)

        # Search filter
        if q:
            search_term = f"%{q}%"
            query = query.filter(
                (Ad.title.ilike(search_term)) |
                (Ad.description.ilike(search_term))
            )

        # Category filter
        if category:
            query = query.filter(Ad.category_slug == category)

        # Sorting
        if sort == "views":
            query = query.order_by(desc(Ad.views_count), desc(Ad.created_at))
        elif sort == "favs":
            query = query.order_by(desc(Ad.favorites_count), desc(Ad.created_at))
        elif sort == "ai":
            query = query.order_by(desc(Ad.likes_count), desc(Ad.views_count), desc(Ad.created_at))
        else:  # date
            query = query.order_by(desc(Ad.created_at))

        # Get total count
        total = query.count()

        # Pagination
        offset = (page - 1) * pageSize
        ads = query.offset(offset).limit(pageSize).all()

        # Format ads for frontend
        formatted_ads = []
        for ad in ads:
            # Build media URLs
            media_url = ""
            media_files = []

            if ad.media_path:
                media_url = f"/company/data/{ad.media_path}"

            if ad.media_filename:
                # Check if it's a JSON array or single file
                if ad.media_filename.startswith('['):
                    try:
                        files = json.loads(ad.media_filename)
                    except:
                        files = [ad.media_filename]
                else:
                    files = [ad.media_filename]

                media_files = [
                    f"/company/data/{ad.category_slug}/{ad.company_slug}/{ad.ad_id}/{f}"
                    for f in files
                ]

            # Get category and company names
            cat = db.query(Category).filter(Category.category_slug == ad.category_slug).first()
            comp = db.query(Company).filter(Company.company_slug == ad.company_slug).first()

            formatted_ads.append({
                "ad_id": ad.ad_id,
                "title": ad.title,
                "description": ad.description or "",
                "category": ad.category_slug,
                "category_name": cat.category_name if cat else ad.category_slug.title(),
                "company": ad.company_slug,
                "company_name": comp.company_name if comp else ad.company_slug.title(),
                "media": media_files[0] if media_files else media_url,
                "media_files": media_files,
                "media_path": media_url,
                "media_type": ad.media_type or "image",
                "timestamp": str(ad.created_at) if ad.created_at else "",
                "views": ad.views_count or 0,
                "likes": ad.likes_count or 0,
                "favorites": ad.favorites_count or 0,
                "contacts": ad.contacts_count or 0,
                "contact": {
                    "phone": ad.contact_phone or "",
                    "sms": ad.contact_sms or "",
                    "email": ad.contact_email or "",
                    "whatsapp": ad.contact_whatsapp or ""
                }
            })

        return {
            "success": True,
            "ads": formatted_ads,
            "page": page,
            "pageSize": pageSize,
            "total": total,
            "totalPages": (total + pageSize - 1) // pageSize
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={
                "success": False,
                "error": "Failed to fetch ads",
                "message": str(e)
            }
        )


@router.get("/ads/{ad_id}")
async def get_single_ad(ad_id: str, db: Session = Depends(get_db)):
    """Fetch single ad by ID"""
    try:
        ad = db.query(Ad).filter(Ad.ad_id == ad_id).first()

        if not ad:
            return JSONResponse(
                status_code=404,
                content={"success": False, "error": "Ad not found"}
            )

        media_url = f"/company/data/{ad.media_path}" if ad.media_path else ""

        return {
            "success": True,
            "ad": {
                "ad_id": ad.ad_id,
                "title": ad.title,
                "description": ad.description or "",
                "category": ad.category_slug,
                "company": ad.company_slug,
                "media_path": media_url,
                "media_type": ad.media_type or "image",
                "views": ad.views_count or 0,
                "likes": ad.likes_count or 0,
                "favorites": ad.favorites_count or 0,
                "contacts": ad.contacts_count or 0,
                "contact": {
                    "phone": ad.contact_phone or "",
                    "sms": ad.contact_sms or "",
                    "email": ad.contact_email or "",
                    "whatsapp": ad.contact_whatsapp or ""
                },
                "created_at": str(ad.created_at) if ad.created_at else ""
            }
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )

File: python_system/python_shared/api/get_ads.php
<?php
/********************************************
 * get_ads.php - Hybrid Database System
 * Fetches ads from SQLite database
 ********************************************/

// Suppress all warnings and notices
error_reporting(E_ERROR | E_PARSE);
ini_set('display_errors', '0');

// Start output buffering to catch any stray output
ob_start();

header("Content-Type: application/json");
header("Access-Control-Allow-Origin: *");

// Load database system
require_once __DIR__ . '/../shared/database/Database.php';
require_once __DIR__ . '/../shared/database/AdModel.php';

$db = Database::getInstance();
$adModel = new AdModel();

// Get parameters
$page = max(1, intval($_GET["page"] ?? 1));
$q = trim($_GET["q"] ?? "");
$category = trim($_GET["category"] ?? "");
$company = trim($_GET["company"] ?? ""); // Add company filter
$sort = $_GET["sort"] ?? "date";
$pageSize = 12;

try {
    // Build SQL query
    $sql = "SELECT
                a.*,
                c.category_name,
                comp.company_name
            FROM ads a
            LEFT JOIN categories c ON a.category_slug = c.category_slug
            LEFT JOIN companies comp ON a.company_slug = comp.company_slug
            WHERE a.status = 'active'";

    $params = [];

    // Company filter (for my_ads.php and dashboard)
    if (!empty($company)) {
        $sql .= " AND a.company_slug = ?";
        $params[] = $company;
    }

    // Search filter
    if (!empty($q)) {
        $sql .= " AND (a.title LIKE ? OR a.description LIKE ?)";
        $searchTerm = "%$q%";
        $params[] = $searchTerm;
        $params[] = $searchTerm;
    }

    // Category filter
    if (!empty($category)) {
        $sql .= " AND a.category_slug = ?";
        $params[] = $category;
    }

    // Sorting
    switch ($sort) {
        case 'views':
            $sql .= " ORDER BY a.views_count DESC, a.created_at DESC";
            break;
        case 'favs':
            $sql .= " ORDER BY a.favorites_count DESC, a.created_at DESC";
            break;
        case 'ai':
            $sql .= " ORDER BY a.likes_count DESC, a.views_count DESC, a.created_at DESC";
            break;
        case 'date':
        default:
            $sql .= " ORDER BY a.created_at DESC";
            break;
    }

    // Get total count
    $countSql = str_replace('SELECT a.*, c.category_name, comp.company_name', 'SELECT COUNT(*) as total',
                           substr($sql, 0, strpos($sql, 'ORDER BY') ?: strlen($sql)));

    $totalResult = $db->queryOne($countSql, $params);
    $total = $totalResult['total'] ?? 0;

    // Pagination
    $offset = ($page - 1) * $pageSize;
    $sql .= " LIMIT ? OFFSET ?";
    $params[] = $pageSize;
    $params[] = $offset;

    // Execute query
    $ads = $db->query($sql, $params);

    // Format ads for frontend
    $formattedAds = [];
    foreach ($ads as $ad) {
        // Build media path
        $mediaPath = "/services/company/data/" . $ad['media_path'];

        // Parse media files (if multiple)
        $mediaFiles = [];
        if (!empty($ad['media_filename'])) {
            // Check if it's a JSON array or single file
            if (strpos($ad['media_filename'], '[') === 0) {
                $mediaFiles = json_decode($ad['media_filename'], true) ?? [$ad['media_filename']];
            } else {
                $mediaFiles = [$ad['media_filename']];
            }
        }

        // Build full media URLs
        $mediaUrls = array_map(function($file) use ($ad) {
            return "/services/company/data/{$ad['category_slug']}/{$ad['company_slug']}/{$ad['ad_id']}/$file";
        }, $mediaFiles);

        $formattedAds[] = [
            'ad_id' => $ad['ad_id'],
            'title' => $ad['title'],
            'description' => $ad['description'],
            'category' => $ad['category_slug'],
            'category_name' => $ad['category_name'] ?? ucfirst($ad['category_slug']),
            'company' => $ad['company_slug'],
            'company_name' => $ad['company_name'] ?? ucfirst($ad['company_slug']),
            'media' => $mediaUrls[0] ?? '', // Primary media
            'media_files' => $mediaUrls, // All media files
            'media_type' => $ad['media_type'] ?? 'image',
            'timestamp' => $ad['created_at'],
            'views' => (int)($ad['views_count'] ?? 0),
            'likes' => (int)($ad['likes_count'] ?? 0),
            'favorites' => (int)($ad['favorites_count'] ?? 0),
            'contacts' => (int)($ad['contacts_count'] ?? 0),
            'contact' => [
                'phone' => $ad['contact_phone'] ?? '',
                'sms' => $ad['contact_sms'] ?? '',
                'email' => $ad['contact_email'] ?? '',
                'whatsapp' => $ad['contact_whatsapp'] ?? ''
            ]
        ];
    }

    // Return response
    $response = [
        'success' => true,
        'ads' => $formattedAds,
        'page' => $page,
        'pageSize' => $pageSize,
        'total' => $total,
        'totalPages' => ceil($total / $pageSize)
    ];

} catch (Exception $e) {
    http_response_code(500);
    $response = [
        'success' => false,
        'error' => 'Failed to fetch ads',
        'message' => $e->getMessage()
    ];
}

// Clean output buffer
ob_end_clean();

// Output JSON
echo json_encode($response);
exit;
